sn_node/src/put_validation.rs

```diff
@@ -503,10 +503,24 @@ impl Node {
 
         // unpack transfer
         trace!("Unpacking incoming Transfers for record {pretty_key}");
-        let (received_fee, cash_notes, royalties_cash_notes_r) = self
+        let (received_fee, mut cash_notes, royalties_cash_notes_r) = self
             .cash_notes_from_transfers(payment.transfers, &wallet, pretty_key.clone())
             .await?;
 
+        // check for spends that we have already received
+        cash_notes.retain(|cash_note| {
+            let spend_addr = SpendAddress::from_unique_pubkey(&cash_note.unique_pubkey());
+            let double_spend = matches!(wallet.get_confirmed_spend(spend_addr), Ok(Some(_spend)));
+            if double_spend {
+                warn!(
+                    "Double spend {} detected for record payment {pretty_key}",
+                    cash_note.unique_pubkey()
+                );
+            }
+            // retain the `CashNote` if it's not a double spend
+            !double_spend
+        });
+
```

Comment:
ah, if `cash_notes` becomes `empty` after the filter,
then there is actually no point to undertake the following steps.
so we shall have something like
```rust
if cash_notes.is_empty() {
    info!("All incoming cash_notes are already processed, no need for the further process!");
    return Err(Error::ReusedPayment);
}
```